### PR TITLE
Implement vmsumcud and vmsumudm instructions

### PR DESCRIPTION
**Pull request by vcoracolombo**
Inline implementations of vmsumcud and vmsumudm, based on https://lists.gnu.org/archive/html/qemu-ppc/2020-07/msg00419.html, a patch sent to the list but never merged.

### CODE REVIEW

**Review by mferst — COMMENTED** [review]
The code is looking good, just some nitpicking about names/style.

File: target/ppc/translate/vmx-impl.c.inc [hunk 1]
```diff
@@ -2104,6 +2104,42 @@ static bool trans_VPEXTD(DisasContext *ctx, arg_VX *a)
...
+
+    get_avr64(src1, a->vra, 0);
+    get_avr64(src2, a->vrb, 0);
+    tcg_gen_mulu2_i64(src1, src2, src1, src2);
+    tcg_gen_add2_i64(rl, rh, rl, rh, src1, src2);
+
+    get_avr64(src1, a->vra, 1);
+    get_avr64(src2, a->vrb, 1);
+    tcg_gen_mulu2_i64(src1, src2, src1, src2);
+    tcg_gen_add2_i64(rl, rh, rl, rh, src1, src2);
```
[the remark this answers is not in the document]
Comment (vcoracolombo) [reply target/ppc/translate/vmx-impl.c.inc]: Nice, thanks for the idea. Changed in 2a4ca49

File: target/ppc/translate/vmx-impl.c.inc [hunk 2]
```diff
@@ -2104,6 +2104,49 @@ static bool trans_VPEXTD(DisasContext *ctx, arg_VX *a)
...
+
+    rh = tcg_temp_new_i64();
+    rl = tcg_temp_new_i64();
+    tmp1h = tcg_temp_new_i64();
+    tmp1l = tcg_temp_new_i64();
+    tmp0h = tcg_temp_new_i64();
+    tmp0l = tcg_temp_new_i64();
+    zero = tcg_constant_i64(0);
+
+    get_avr64(rh, a->vra, 0);
```
[the remark this answers is not in the document]
Comment (vcoracolombo) [reply target/ppc/translate/vmx-impl.c.inc]: Fixed in 3eddea2, thanks!

File: target/ppc/translate/vmx-impl.c.inc [hunk 3]
```diff
@@ -2104,6 +2104,49 @@ static bool trans_VPEXTD(DisasContext *ctx, arg_VX *a)
...
+    get_avr64(rl, a->vrb, 0);
+    tcg_gen_mulu2_i64(tmp1l, tmp1h, rh, rl);
+    get_avr64(rh, a->vra, 1);
+    get_avr64(rl, a->vrb, 1);
```
[the remark this answers is not in the document]
Comment (vcoracolombo) [reply target/ppc/translate/vmx-impl.c.inc]: Fixed in 3eddea2. Unfortunately I couldn't think of a better name for `rh`/`rl`, so I changed it to `tmp0`/`tmp1` as you suggested. If you have a better idea please let me know.

File: target/ppc/translate/vmx-impl.c.inc [hunk 4]
```diff
@@ -2104,6 +2104,49 @@ static bool trans_VPEXTD(DisasContext *ctx, arg_VX *a)
     return true;
 }
 
+static bool trans_VMSUMCUD(DisasContext *ctx, arg_VA *a)
+{
+    TCGv_i64 rh, rl, tmp1h, tmp1l, tmp0h, tmp0l, zero;
```
[the remark this answers is not in the document]
Comment (vcoracolombo) [reply target/ppc/translate/vmx-impl.c.inc]: Fixed in 3eddea2, thanks for the suggestion!

[the remark this answers is not in the document]
Comment (mferst) [reply target/ppc/translate/vmx-impl.c.inc]: : it seems that double word zero is used to calculate `prod1` and double word one is used to calculate `prod0`.

**Review by mferst — APPROVED** [review]
It would be nice to have the comments (like the ones in the original patch) in VMSUMCUD, but it's fine without them too.

File: target/ppc/translate/vmx-impl.c.inc (hunk 4 above)
[the remark this answers is not in the document]
Comment (vcoracolombo) [reply target/ppc/translate/vmx-impl.c.inc]: you mean in the following part?

```c
get_avr64(tmp0, a->vra, false);
get_avr64(tmp1, a->vrb, false);
tcg_gen_mulu2_i64(prod1l, prod1h, tmp0, tmp1);
get_avr64(tmp0, a->vra, true);
get_avr64(tmp1, a->vrb, true);
tcg_gen_mulu2_i64(prod0l, prod0h, tmp0, tmp1);
```

I think this is correct. double word zero is the leftmost (most significant) part, and double word one is the rightmost (least significant). when we do `get_avr64(tmp0, a->vra, false);` we are getting double word 1, and multiplying with `tcg_gen_mulu2_i64(prod1l, prod1h, tmp0, tmp1);`, which is storing in `prod1`, right?

[the remark this answers is not in the document]
Comment (mferst) [reply target/ppc/translate/vmx-impl.c.inc]: Yep, that's right, I got it wrong

**Comment by vcoracolombo** [comment]
> It would be nice to have the comments (like the ones in the original patch) in VMSUMCUD, but it's fine without them too.

Added on 18021b0.

**Comment by mferst** [comment]
Merged in upstream: https://lists.gnu.org/archive/html/qemu-ppc/2022-03/msg00002.html

**Pull request closed without merging**